consensus/src/block_storage/block_tree.rs

```diff
@@ -225,6 +225,9 @@ where
         // qc1 == qc2 || qc1.round != qc2.round
         // The invariant is quadratic but can be maintained in linear time by the check
         // below.
+
+        // @REVIEW: If this is true, is there a reason this data structure isn't indexed by round?
+        // Also, What if this invariant is violated (e.g., due to byzantine cases? what would happen?)
```

Comment:
> if there's more than 1 QC in the same round, safety is violated. what would happen? well this chain is doomed we need writeset and blabla.

My point is is more: is the behavior when we run into this situation "panic and inform the system operator". Have we stored enough data to disk at this point to allow the operator to determine the exact block that caused this issue?

> re indexing, it's indexed by id so that we can easily query by block id, there can be multiple blocks in the same round.

But only one block per round can have a QC formed on it (and a block without a QC can never be committed). So it seems like we really only have to remember one block per round.

Reply:
we only have to remember one block per round eventually, but when there's multiple proposals coming, you have to remember both until one of them gets QC

Reply:
When does the leader for the next round fetch blocks — after they get 2f+1 votes, or does it fetch every block it sees votes on?

Reply:
nope, it only fetches block after QC aggregated. you want to propose we only store the block we voted on and fetch blocks that have QC? I think that's probably doable

Reply:
at least from my read that could substantially simplify assumptions (e.g., you could index by round -> you can use ranged deletes to GC). I don't think you have to do this.... but it seems like an interesting approach to consider